### PR TITLE
feat: prefill email in login form after changing email address

### LINKED ISSUE #3140: Prefill the login page email address after changing email
Just a nice improvement

### PR DESCRIPTION
**Pull request by kacperknapik**
Closes #3140

### CODE REVIEW

**Comment by kacperknapik** [comment]
I'm worried about show the email in the URL. reviewer should take a look a help me decide

**Comment by marlass** [comment]
You have an option to use `history.state` if you don't want it visible.

```
this.routingService.go({ cxRoute: 'login' }, { }, {
    state: {
      mail: this.newUid
    }
});

// in login component
history.state // { mail: 'passed email' }
```

File: projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts [hunk 1]
```diff
@@ -8,40 +8,60 @@ import {
...
+  subs: Subscription;
   form: FormGroup;
 
   constructor(
     private auth: AuthService,
     private globalMessageService: GlobalMessageService,
     private fb: FormBuilder,
-    private authRedirectService: AuthRedirectService
+    private authRedirectService: AuthRedirectService,
+    private route: ActivatedRoute
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts]: That is a breaking change. Let's make this service optional and if this service will be provided then support email autofill.

File: projects/storefrontlib/src/cms-components/user/login-form/login-form.component.spec.ts [hunk 2]
```diff
@@ -41,12 +42,14 @@ describe('LoginFormComponent', () => {
 
   let authService: MockAuthService;
   let authRedirectService: AuthRedirectService;
+  let mockWindowRef: WindowRef;
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/user/login-form/login-form.component.spec.ts]: rename it to winRef as it is not mocked

File: projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts [hunk 3]
```diff
@@ -14,48 +15,66 @@ import { CustomFormValidators } from '../../../shared/utils/validators/custom-fo
...
+  subs: Subscription;
   form: FormGroup;
 
   constructor(
     private auth: AuthService,
     private globalMessageService: GlobalMessageService,
     private fb: FormBuilder,
-    private authRedirectService: AuthRedirectService
+    private authRedirectService: AuthRedirectService,
+    @Optional() private winRef: WindowRef
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts]: My mistake here. It should be `private winRef?: WindowsRef`.

File: projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts [hunk 4]
```diff
@@ -14,48 +15,66 @@ import { CustomFormValidators } from '../../../shared/utils/validators/custom-fo
...
-  ngOnInit() {
+  ngOnInit(): void {
     this.form = this.fb.group({
       userId: ['', [Validators.required, CustomFormValidators.emailValidator]],
       password: ['', Validators.required],
     });
+
+    if (this.winRef) {
+      const routeState =
+        this.winRef.nativeWindow.history &&
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts]: Add also check for `this.winRef.nativeWindow` it can be `undefined` in SSR

File: projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts [hunk 5]
```diff
@@ -14,48 +15,66 @@ import { CustomFormValidators } from '../../../shared/utils/validators/custom-fo
...
-        }
-      });
+    const { userId, password } = this.form.controls;
+    this.auth.authorize(
+      userId.value.toLowerCase(), // backend accepts lowercase emails only
+      password.value
+    );
+
+    if (!this.tokenExists) {
+      this.subs = new Subscription().add(
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts]: There is only one subscription in the component. No need to use `.add` method.

File: projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts (hunk 3 above)
Comment (marlass) [reply projects/storefrontlib/src/cms-components/user/login-form/login-form.component.ts]: Please provide deprecation comment to the constructor to make `winRef` required in later releases.

File: .editorconfig [hunk 6]
```diff
@@ -2,6 +2,7 @@
 root = true
 
 [*]
+end_of_line = lf
```
Comment (marlass) [inline .editorconfig]: Could you move it to separate PR?

**Comment by kacperknapik** [comment]
Thanks @WTPar for your contribution :)
